File: examples/csharp/minimal-order-service/Domain/Order.cs
namespace OrderService.Domain;

/// <summary>
/// Domain entity representing an order.
/// </summary>
public class Order
{
    private OrderStatus _status;

    public Order(string id, Money total)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("Order ID cannot be null or empty", nameof(id));
        }
        if (total == null)
        {
            throw new ArgumentNullException(nameof(total));
        }
        
        Id = id;
        Total = total;
        _status = OrderStatus.Pending;
    }

    public static Order Create(Money total)
    {
        var id = Guid.NewGuid().ToString();
        return new Order(id, total);
    }

    public string Id { get; }
    public Money Total { get; }
    public OrderStatus Status => _status;

    public void MarkAsPaid()
    {
        if (_status != OrderStatus.Pending)
        {
            throw new InvalidOperationException("Only pending orders can be marked as paid");
        }
        _status = OrderStatus.Paid;
    }

    public override bool Equals(object? obj)
    {
        if (obj is Order other)
        {
            return Id == other.Id;
        }
        return false;
    }

    public override int GetHashCode()
    {
        return Id.GetHashCode();
    }

    public override string ToString()
    {
        return $"Order{{Id='{Id}', Total={Total}, Status={Status}}}";
    }
}

public enum OrderStatus
{
    Pending,
    Paid
}

File: examples/cpp/minimal-order-service/domain/order.cpp
#include "domain/order.hpp"
#include <random>
#include <sstream>
#include <iomanip>

namespace domain {

Order::Order(const std::string& id, std::shared_ptr<Money> total)
    : id_(id), total_(total), status_(OrderStatus::PENDING) {
    if (id.empty()) {
        throw std::invalid_argument("Order ID cannot be empty");
    }
    if (!total) {
        throw std::invalid_argument("Total cannot be null");
    }
}

std::unique_ptr<Order> Order::create(std::shared_ptr<Money> total) {
    std::string id = generateId();
    return std::make_unique<Order>(id, total);
}

void Order::markAsPaid() {
    if (status_ != OrderStatus::PENDING) {
        throw std::logic_error("Only pending orders can be marked as paid");
    }
    status_ = OrderStatus::PAID;
}

std::string Order::generateId() {
    // Simple UUID-like generation
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    
    std::ostringstream oss;
    oss << "order-";
    for (int i = 0; i < 32; ++i) {
        oss << std::hex << dis(gen);
    }
    return oss.str();
}

} // namespace domain
